Airspeed Estimation Module Tested ok in cyberzoo with 3b

frlarocque opened the pull request:
Adding module "ekf_aw" (extended kalman filter for airspeed and wind" to estimate airspeed and wind speed. Module uses GPS velocity, acceleration, euler angles, RPM from the pusher, RPM from the hover props, skew angle and elevator angle.

**Changes:**
1. Module itself
-  ekf_aw.xml
- ekf_aw.cpp
- ekf_aw.h
- ekf_aw_wrapper.c
- ekf_aw_wrapper.h

2. ABI Message for RPM
- actuators.h : new struct for rpm abi message
- abi_common.h : add the header file where the new struct for rpm is defined
- abi.xml : change structure of abi message
- actuators_uavcan.c : send rpm message
- stabilization_indi.c : only place where rpm abi message is used in paparazzi, but not used in rotwing case. Still modified to work with new abi message

3. Telemetry change to add ekf_aw messages
- highspeed_rotorcraft.xml : added 3 ekf_aw messages for telemetry to SD and to ground station

4. Airframe files
- rot_wing_v3a.xml : added module to airframe (with define set for v3b)
- rot_wing_v3b.xml : added module to airframe (with define set for v3a) and new magnetometer calibration for leg mounted magneto (check with Dennis if still required)

**Checks completed:**
- Module does not use dynamically allocated matrices ([using PX4 matrix module running in EKF2](https://github.com/PX4/PX4-Autopilot/tree/4a3d64f1d76856d22323d1061ac6e560efda0a05/src/lib/matrix))
- Module runs on ground on v3a in approx. 140 micro sec
- Module runs on ground on v3b in approx. 140 micro sec
- Flight test in Cyberzoo with v3b for 50 s (repeated 2 times)
- CPU load in flight with module approx. 10% (from RTOS telemetry message, supposed to be +-5% precise). Unfortunately no comparison to before as RTOS message was not logged before

Commit:
Airspeed Estimation Module Tested ok in cyberzoo with 3b

frlarocque committed:
Merge branch 'rotwing_master_3b' into airspeed_wind_module_3b

EwoudSmeur commented on sw/airborne/modules/actuators/actuators_uavcan.c:
```diff
@@ -257,6 +268,7 @@ void actuators_uavcan_commit(struct uavcan_iface_t *iface, int16_t *values, uint
   // Broadcast the raw command message on the interface
   uavcan_broadcast(iface, UAVCAN_EQUIPMENT_ESC_RAWCOMMAND_SIGNATURE, UAVCAN_EQUIPMENT_ESC_RAWCOMMAND_ID,
                    CANARD_TRANSFER_PRIORITY_HIGH, buffer, (offset + 7) / 8);
+
```
Please remove this whitespace

frlarocque replied:
Adressed in commit 85dceb661fcbabb175ae82984f055dd64fdd3bd1

EwoudSmeur commented on conf/airframes/tudelft/rot_wing_v3a.xml:
```diff
@@ -128,6 +128,64 @@
+        
+            <define name="EKF_AW_WRAPPER_ROT_WING_TYPE_A"  value="1" description="1 for A, 0 for b"/>
+
+            <define name="EKF_AW_P0_V_body"  value="1.E-2f" description="Initial covariance body velocity"/>
```
defines should be all caps

frlarocque replied:
Adressed in commit 85dceb661fcbabb175ae82984f055dd64fdd3bd1

EwoudSmeur commented on sw/airborne/modules/meteo/ekf_aw.cpp:
```diff
@@ -0,0 +1,1489 @@
+    // X Axis
+    ekf_aw_params.k_fx_drag[0] = EKF_AW_K1_FX_DRAG; ekf_aw_params.k_fx_drag[1] = EKF_AW_K2_FX_DRAG;
+    ekf_aw_params.k_fx_fuselage[0] = EKF_AW_K1_FX_FUSELAGE; ekf_aw_params.k_fx_fuselage[1] = EKF_AW_K2_FX_FUSELAGE; ekf_aw_params.k_fx_fuselage[2] = EKF_AW_K3_FX_FUSELAGE; ekf_aw_params.k_fx_fuselage[3] = EKF_AW_K4_FX_FUSELAGE;
+    ekf_aw_params.k_fx_hover[0] = EKF_AW_K1_FX_HOVER; ekf_aw_params.k_fx_hover[1] = EKF_AW_K2_FX_HOVER;ekf_aw_params.k_fx_hover[2] = EKF_AW_K3_FX_HOVER;
```
Please don't make the lines so long

frlarocque replied:
Adressed in commit 85dceb661fcbabb175ae82984f055dd64fdd3bd1

EwoudSmeur commented on sw/airborne/modules/meteo/ekf_aw.cpp:
```diff
@@ -0,0 +1,1489 @@
+  dcm(2, 2) = cos_phi * cos_theta;
+
+  // Calculate Angle of Attack
+  float aoa = atan2(w, u==0 ? 1E-5 : u); // Protected alpha calculation
```
atan2f is in float, atan2 is in double precision. The atan2 function should not have any singularity (maybe check for (0,0)

frlarocque replied:
corrected to atanf in commit 85dceb661fcbabb175ae82984f055dd64fdd3bd1
Should indeed not have any singularity according to https://en.cppreference.com/w/c/numeric/math/atan2

EwoudSmeur commented on sw/airborne/modules/meteo/ekf_aw.cpp:
```diff
@@ -0,0 +1,1489 @@
+
+  // Calculate Angle of Attack
+  float aoa = atan2(w, u==0 ? 1E-5 : u); // Protected alpha calculation
+  float sat_aoa = aoa < deg2rad*EKF_AW_AOA_MIN_ANGLE ? deg2rad*EKF_AW_AOA_MIN_ANGLE : aoa > deg2rad*EKF_AW_AOA_MAX_ANGLE ? deg2rad*EKF_AW_AOA_MAX_ANGLE : aoa; // Saturate alpha
```
PPRZ has a 'Bound' function for this

frlarocque replied:
Adressed in commit 85dceb661fcbabb175ae82984f055dd64fdd3bd1

EwoudSmeur commented on sw/airborne/modules/meteo/ekf_aw.cpp:
```diff
@@ -0,0 +1,1489 @@
+  float diff_alpha_u = 0;
+  float diff_alpha_w = 0;
+  if (u > 1E-5){
+    diff_alpha_u = -w/(u*u*(tan_aoa*tan_aoa + 1)); // -w/(u^2*(w^2/u^2 + 1))
```
What if tan_aoa==-1?

EwoudSmeur commented on sw/airborne/modules/meteo/ekf_aw.cpp:
```diff
@@ -0,0 +1,1489 @@
+  // Calculate Sideslip
+  float beta = 0;
+  if (V_a > 1E-5){
+    beta = asinf(v/V_a < -1 ? -1 : v/V_a > 1 ? 1 : v/V_a); // Ratio is kept between -1 and 1
```
use 1.0f for floating point numbers

EwoudSmeur commented on sw/airborne/modules/meteo/ekf_aw.cpp:
```diff
@@ -0,0 +1,1489 @@
+    exp_az = 0;
+  }
+  else{
+    exp_az = EKF_AW_AZ_SCHED_GAIN+(eawp.inputs.skew-deg2rad*EKF_AW_AZ_SCHED_START_DEG)*((EKF_AW_AZ_SCHED_GAIN)/(deg2rad*(EKF_AW_AZ_SCHED_START_DEG-EKF_AW_AZ_SCHED_END_DEG)));
```
start_deg can never be equal to end_deg!

EwoudSmeur commented on sw/airborne/modules/meteo/ekf_aw.cpp:
```diff
@@ -0,0 +1,1489 @@
+    G(3,2) += (2*w*(temp_1) + (ekf_aw_params.k_fx_fuselage[2]*diff_alpha_w + 2*ekf_aw_params.k_fx_fuselage[3]*aoa*diff_alpha_w)*V_a*V_a)/ekf_aw_params.vehicle_mass;
+    
+    // Hover prop contribution
+    G(3,0) += (2*ekf_aw_params.k_fx_hover[0]*sign_u*u + (0.5*ekf_aw_params.k_fx_hover[1]*sign_u*hover_RPM_mean*hover_RPM_mean)/sqrt(abs(u ==0 ? 1E-5 : u))+ ekf_aw_params.k_fx_hover[2] * sign_u)/ekf_aw_params.vehicle_mass; // TO DO: Verify the abs()
```
use sqrtf, fabsf

frlarocque replied:
Adressed in commit 85dceb661fcbabb175ae82984f055dd64fdd3bd1

EwoudSmeur commented on sw/airborne/modules/meteo/ekf_aw.cpp:
```diff
@@ -0,0 +1,1489 @@
+  // A_y_filt related lines
+  if (EKF_AW_USE_BETA){
+    float protected_V_a = V_a ==0 ? 1E-5 : V_a;
+    float cos_beta = cosf(beta) == 0 ? 1E-5 : cosf(beta);
```
cos(beta) and or V_a can still become very very small, which can lead to inf in the division. It would be better to check if abs(V_a) > 1e-5

frlarocque replied:
I changed  all occurences val==other_val by fabsf(val)<other_val.
Adressed in commit 85dceb661fcbabb175ae82984f055dd64fdd3bd1

EwoudSmeur commented on sw/airborne/modules/meteo/ekf_aw.cpp:
```diff
@@ -0,0 +1,1489 @@
+  float psi = eawp.inputs.euler(2);
+
+  // FOR DEBUG
+  if (EKF_AW_DEBUG){
```
In all places below:

EKF_AW_DEBUG is a define, which you can check at compile time with
#if EKF_AW_DEBUG
code
#endif

'code' will not be compiled if the flag is false

frlarocque replied:
Adressed in commit 85dceb661fcbabb175ae82984f055dd64fdd3bd1

EwoudSmeur commented on sw/airborne/modules/meteo/ekf_aw.cpp:
```diff
@@ -0,0 +1,1489 @@
+{
+  float diagonal[EKF_AW_R_SIZE];
+  for(int8_t i=0; i<EKF_AW_R_SIZE; i++) {
+    diagonal[i] = log10(eawp.R(i,i));
```
what if R is zero?

frlarocque replied:
Good catch! I protected against values smaller or equal to zero in commit 85dceb661fcbabb175ae82984f055dd64fdd3bd1

EwoudSmeur commented on sw/airborne/modules/meteo/ekf_aw.cpp:
```diff
@@ -0,0 +1,1489 @@
+{
+  float diagonal[EKF_AW_COV_SIZE];
+  for(int8_t i=0; i<EKF_AW_COV_SIZE; i++) {
+    diagonal[i] = log10(eawp.P(i,i));
```
same here, log10 if R is zero?
and below

frlarocque replied:
Good catch! I protected against values smaller or equal to zero in commit 85dceb661fcbabb175ae82984f055dd64fdd3bd1

EwoudSmeur reviewed (changes requested):
Looks good, can you please address the comments?

Commit:
Addressed pull request 92 comments

Removed whitespace in actuators_uavcan.c, All define are in MAJ, put long lines in separate lines, corrected atan2() for atan2f(), used Bound() function for saturation, protected derivative of alpha against 0,made sure all floating point numbers were defined as X.Xf, protected covariance scheduling against same start and end, changed sqrt() to sqrtf, changed abs() to fabsf(), instead of  val == 0 used fabsf(val)<1E-5, changed all if/else statements using define to #if #else #endif, protected log10 against negative values and 0

frlarocque commented:
Adressed comments in commit 85dceb661fcbabb175ae82984f055dd64fdd3bd1

Was able to build for v3a and v3b
